async_service/trio.py

```diff
@@ -17,40 +18,72 @@
 import trio
 import trio_typing
 
-from ._utils import get_task_name
+from ._utils import get_task_name, iter_dag
 from .abc import ManagerAPI, ServiceAPI
 from .base import BaseManager
 from .exceptions import DaemonTaskExit, LifecycleError, ServiceCancelled
 from .typing import EXC_INFO
 
 
 class TrioManager(BaseManager):
+    _is_cancelled = False
+
     # A nursery for sub tasks and services.  This nursery is cancelled if the
     # service is cancelled but allowed to exit normally if the service exits.
     _task_nursery: trio_typing.Nursery
 
+    _service_task_dag: Dict[trio.hazmat.Task, List[trio.hazmat.Task]]
+    _task_done_events: Dict[trio.hazmat.Task, trio.Event]
+    _task_cancel_scopes: Dict[trio.hazmat.Task, trio.CancelScope]
+
     def __init__(self, service: ServiceAPI) -> None:
         super().__init__(service)
 
         # events
         self._started = trio.Event()
+        self._cancelled = trio.Event()
         self._stopping = trio.Event()
         self._finished = trio.Event()
 
         # locks
         self._run_lock = trio.Lock()
 
+        # DAG tracking
+        self._service_task_dag = {}
+        self._task_cancel_scopes = {}
+        self._task_done_events = {}
+
     #
     # System Tasks
     #
+    async def _handle_cancelled(self) -> None:
+        # Here we iterate over the task DAG such that we cancel the most deeply
+        # nested tasks first ensuring that each has finished completely before
+        # we move onto cancelling the parent tasks.
+        self.logger.debug("%s: _handle_cancelled waiting for cancellation", self)
+        await self._cancelled.wait()
+        self.logger.debug("%s: _handle_cancelled triggering task cancellation", self)
+
+        for task in iter_dag(self._service_task_dag):
+            cancel_scope = self._task_cancel_scopes[task]
+            done = self._task_done_events[task]
+            cancel_scope.cancel()
+            await done.wait()
```

Comment:
One of the trio philosophies is to not bake timeouts into your APIs.  I'll happily re-visit this if it proves a problem but right now I'm not sold on it being the right approach.

Reply:
I didn't mean to add a timeout parameter anywhere, but ISTM that if we have a rogue task that doesn't return, it would prevent the done event from being set and the service would hang forever here. My suggestion was to wrap the `await done.wait()` with a `trio.move_on_after(hardcoded_timeout)` or something to that effect

Reply:
I think if this hangs then it means that someone wrote something like this in one of their tasks.

```python
try:
    ...
finally:
    with trio.CancelScope(shield=True):
        ...  # do cleanup....
```

This means they have explicitly blocked shutdown which suggests that hanging is the right choice.

Reply:
Or maybe a task that catches a Cancelled exception and fails to re-raise it?  My thinking was that it would probably be a bug in the task itself and instead of hanging we could log a warning but still terminate the service. Otherwise it could be hard to find out what was causing the service to hang